### PR TITLE
Bug/fix sessions load

### PR DESCRIPTION
**Pull request by Secretmapper**
React-transmission doesn't read/save to cookies necessary data (compact, etc.). Sort of bundled up #34 here since I realized react-transmission at its current state does not support this.

NOTE: used master as a starting point, which re-introduces some bugs already fixed in other pull requests (iirc https://github.com/fcsonline/react-transmission/pull/33).

### CODE REVIEW

**Comment by fcsonline** [comment]
All preferences should be save in server as we were talking before. The rest of preferences like UI ones could be stored in the client but I think that cookies are not the best approach. IMHO localstorage is a better approach.

**Comment by enricostano** [comment]
Completely agree with @fcsonline here, if we need to store data on client side localstorage is a good option.

File: src/stores/torrent-store.js [hunk 1]
```diff
@@ -22,13 +22,11 @@ const extractDomains = (torrent) => {
...
   @observable previousTorrents = [];
   @observable torrents = [];
-  @observable statusFilter = 0;
   @observable trackerFilter = '';
   @observable textFilter = '';
-  @observable sortCriteria = 'name';
-  @observable sortDirection = '';
 
-  constructor(rpc) {
+  constructor(rpc, prefs_store) {
```
Comment (edulan) [inline src/stores/torrent-store.js]: Here we are coupling stores and this is one of the things we hardly try not to do. Instead it's preferable to aggregate data from them inside components.

File: src/stores/prefs-store.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    this.compact = this.rehydrateKey(PrefCookieKeys.compact, false);
+  }
+
+  @action setStatusFilter(statusFilter) {
+    this.statusFilter = statusFilter;
+
+    // status filter is saved as a string (i.e. 'all')
+    // so we need to make sure to save and read it as so.
+    const persistAs = findByProperty(FilterStates, 'value', statusFilter).persistKey;
+    this.persistKey(PrefCookieKeys.statusFilter, persistAs);
```
Comment (edulan) [inline src/stores/prefs-store.js]: Instead of performing data serialization here, it's much preferable to move this to a [reaction](https://mobx.js.org/refguide/reaction.html).

You can check notifications example in `reactions` folder.

File: src/stores/prefs-store.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+];
+
+class PrefsStore {
+  @observable statusFilter;
+  @observable sortCriteria;
+  @observable sortDirection;
+  @observable compact;
+
+  constructor() {
+    this.rehydrate();
```
Comment (edulan) [inline src/stores/prefs-store.js]: It'd be better to extract store initialization to a different module, like a bootstrap process or similar.

**Comment by Secretmapper** [comment]
Hello! @fcsonline @enricostano 

Note that I'm not saving any server side pref, all of the preferences I'm saving here are client-side only (view options, sort, etc.).

As I noted on https://github.com/fcsonline/react-transmission/issues/34 I agree that localstorage is the best option. I opted to use cookies for feature-parity, because transmission-web uses cookies.

However I guess an even better option is only to initially hydrate the preferences using cookies, but persist them using localstorage, which will allow us to transparently carry over transmission-web preferences while using localstorage in the future. Of course this has the disadvantage of not supporting older browsers, but since we're using screw_ie8, flex, etc., I'm thinking we're supporting modern browsers only anyway.

Let me know what you think!

File: src/stores/torrent-store.js (hunk 1 above)
Comment (Secretmapper) [reply src/stores/torrent-store.js]: This is how I planned on doing this too initially, as this is how I'd do it with ie Relay/Redux,, however I'm not sure of the implications of the computed filteredTorrents - I'm guessing it's an essentially memoized property when a store value is changed. (first foray with MobX)

So I guess my question is just to confirm it'll be best to move the filteredTorrents function to the component @edulan?

File: src/stores/prefs-store.js (hunk 2 above)
Comment (Secretmapper) [reply src/stores/prefs-store.js]: Makes sense!

File: src/stores/torrent-store.js (hunk 1 above)
Comment (Secretmapper) [reply src/stores/torrent-store.js]: Scratch that, I realized I can't actually do that since the computed property is being accessed through multiple components.

While coupling is less than ideal here I'm not sure what we could do. We can just let the components aggregate the data directly but it's non-ideal too IMO as the computed decorator probably helps.

I just checked and it seems like coupling is the accepted solution for this use case in mobx:
https://github.com/mobxjs/mobx/issues/223

Eager to hear if you think there's a better way @edulan!

**Comment by fcsonline** [comment]
After reviewing the original application we realized that they are storing information in cookies as you said and implemented. It's quite mind blowing.

IMO storing information in cookies just to _remember_ what you chose is a bad pattern used years ago. We have `localStorage` alternative that is better and doesn't impact to every single request size we do to the server. In terms of feature parity is good to implement this feature to restore UI selections but not implementing it in the old way is a tradeoff that we can assume becuase we are not breaking the contract between client and server.

On the other hand, if we store/serialize a JSON in some `localStorage` value, we can get rid of all those parsing string types that always are a source of bugs.

**Comment by Secretmapper** [comment]
Hey, slr been busy lately.

Yup that's true enough. I guess when transmission-web was deployed it made sense as no browser supported cookies back then, but at this point I agree that using cookies to support older browsers is detrimental.

I think that using the "read from cookies to hydrate initially then save to localStorage" is the best solution. I'll submit a PR once I have the time!

**Comment by Secretmapper** [comment]
Switched to localstorage, with cookie fallback for non supported browsers

There's a failing test here, cause as I mentioned, this PR reintroduces the pause filter bug and since it's been merged the bug's caught by the test. I'll decouple the stores and add a filteredTorrents selector and let the components aggregate the data.

**Comment by fcsonline** [comment]
`localStorage` is supported for all browsers since 2009. You can check it here: http://caniuse.com/#search=localstorage

I think that we don't need the fallback, remove all the cookie stuff and simplify the code.

**Comment by Secretmapper** [comment]
Good point! (and done)

File: src/util/persistors.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+const genLocalStorageKey = key => `react-transmission.${key}`
+
+export function persistKey (key, val) {
+  window.localStorage.setItem(genLocalStorageKey(key), val);
+}
+
+export function rehydrateKey (key, fallback) {
+  let val = window.localStorage.getItem(genLocalStorageKey(key));
```
Comment (fcsonline) [inline src/util/persistors.js:8]: I think that we can persist the whole object as a JSON string and restore it and don't deal with parse issues

**Review by fcsonline — CHANGES_REQUESTED** [review]

File: src/reactions/persist.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import { reaction } from 'mobx';
+import { findByProperty } from 'util/common';
+import { persistKey } from 'util/persistors';
+import { FilterStates, PrefCookieKeys } from 'stores/prefs-store';
+
+export default function ({ prefs_store }) {
+  const keys = ['statusFilter', 'sortCriteria', 'sortDirection', 'compact']
+  keys.forEach(k => {
+    reaction(
```
Comment (edulan) [inline src/reactions/persist.js]: Here we're creating a different reaction for every key. We can improve that by creating a single reaction that tracks required observables as follows:

```javascript
const prefKeys = ['statusFilter', 'sortCriteria', 'sortDirection', 'compact'];

reaction(
  () => prefKeys.map((prefKey) => ({[prefKey]: prefs_store[prefKey]})),
  prefPair => {
    prefPair.forEach(serializePref);
  }
)
```

File: src/stores/__tests__/prefs-store.spec.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
+import PrefsStore from '../prefs-store'
+
+describe('PrefsStore', () => {
+  describe('hydrates', () => {
+    xit('default values', () => {})
```
Comment (edulan) [inline src/stores/__tests__/prefs-store.spec.js]: If we're not implementing tests at all better remove it :fire:

Comment (Secretmapper) [reply src/stores/__tests__/prefs-store.spec.js]: I disagree with this. xit/pending tests are exactly made for this - to show which edge cases/parts of the code are significant but still untested.

I'm going to be implementing tests for it anyway so I guess it's a non-issue 👍

File: src/reactions/persist.js (hunk 5 above)
Comment (Secretmapper) [reply src/reactions/persist.js]: Nice catch :)

File: src/util/persistors.js (hunk 4 above)
Comment (Secretmapper) [reply src/util/persistors.js:8]: Fixed

File: src/stores/prefs-store.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
+import {observable, action} from 'mobx';
+import Torrent from './torrent';
+
+export const PrefCookieKeys = {
```
Comment (fcsonline) [inline src/stores/prefs-store.js:4]: Could we remove all references to the word _Cookies_, please?

**Comment by fcsonline** [comment]
:recycle: rebase from master please

**Comment by fcsonline** [comment]
Can you rebase this branch, please?

**Comment by fcsonline** [comment]
Closing this PR. If you rebase it, just publish it again
